MSAL/test/unit/native_auth/controllers/MSALNativeAuthMFAControllerTests.swift

```diff
@@ -0,0 +1,408 @@
+        await checkGetAuthMethodsWithIntrospectValidatorError(validatedError: .expiredToken(.init()), expectedType: .generalError)
+        await checkGetAuthMethodsWithIntrospectValidatorError(validatedError: .unexpectedError(.init()), expectedType: .generalError)
+    }
+    
+    func test_whenGetAuthMethods_correctResultShouldBeReturned() async {
+        let expectedContinuationToken = "continuationToken"
+        let internalAuthMethod = MSALNativeAuthInternalAuthenticationMethod(id: "1", challengeType: .oob, challengeChannel: "email", loginHint: "hint")
+        let expectedContext = MSALNativeAuthRequestContext(correlationId: defaultUUID)
+
+        signInRequestProviderMock.mockIntrospectRequestFunc(MSALNativeAuthHTTPRequestMock.prepareMockRequest())
+        signInResponseValidatorMock.introspectValidatedResponse = .authMethodsRetrieved(continuationToken: expectedContinuationToken, authMethods: [internalAuthMethod])
+        
+        let result = await sut.getAuthMethods(continuationToken: expectedContinuationToken, context: expectedContext, scopes: [])
+        result.telemetryUpdate?(.success(()))
+
+        XCTAssertFalse(cacheAccessorMock.validateAndSaveTokensWasCalled)
+        checkTelemetryEventResult(id: .telemetryApiIdMFAGetAuthMethods, isSuccessful: true)
+        if case .selectionRequired(let authMethods, let newState) = result.result {
+            XCTAssertEqual(authMethods.count, 1)
+            XCTAssertEqual(authMethods.first?.challengeType, internalAuthMethod.challengeType.rawValue)
+            XCTAssertEqual(authMethods.first?.id, internalAuthMethod.id)
+            XCTAssertEqual(authMethods.first?.channelTargetType.value, internalAuthMethod.challengeChannel)
+            XCTAssertEqual(authMethods.first?.loginHint, internalAuthMethod.loginHint)
+            XCTAssertEqual(newState.continuationToken, expectedContinuationToken)
+        } else {
+            XCTFail("Expected selectionRequired result")
+        }
+    }
+    
+    func test_whenSubmitChallengeTokenRequestFails_correctErrorShouldBeReturned() async {
+        let expectedContext = MSALNativeAuthRequestContext(correlationId: defaultUUID)
+
+        tokenRequestProviderMock.mockRequestTokenFunc(nil, throwError: MSALNativeAuthError(message: nil, correlationId: defaultUUID))
+        
+        let result = await sut.submitChallenge(challenge: "1234", continuationToken: "CT", context: expectedContext, scopes: [])
+
+        XCTAssertFalse(cacheAccessorMock.validateAndSaveTokensWasCalled)
+        checkTelemetryEventResult(id: .telemetryApiIdMFASubmitChallenge, isSuccessful: false)
+        if case .error(let error, let newState) = result.result {
+            XCTAssertEqual(error.type, .generalError)
+            XCTAssertNotNil(newState)
+        } else {
+            XCTFail("Expected error result")
+        }
+    }
+    
+    func test_whenSubmitChallengeTokenRequestReturnError_correctErrorShouldBeReturned() async {
```

Comment:
I think this test is ok to exists, because it tests a behaviour of our code base. And we're sure that it works as expected.
What I see not optimal in our unit test suite is that a single function is tested in multiple tests. This means that when that function change, many tests (not directly related to that function) needs to be changed.